third_party/amd/include/TritonAMDGPUTransforms/PointerCanonicalizer.h

```diff
@@ -0,0 +1,199 @@
+#ifndef TRITON_DIALECT_TRITONGPU_TRANSFORMS_POINTER_CANONICALIZER_H_
+#define TRITON_DIALECT_TRITONGPU_TRANSFORMS_POINTER_CANONICALIZER_H_
+
+#include "mlir/Dialect/ControlFlow/IR/ControlFlowOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Location.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Value.h"
+#include "mlir/IR/Visitors.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Support/LogicalResult.h"
+#include "triton/Dialect/Triton/IR/Dialect.h"
+
+using namespace mlir;
+
+namespace mlir::triton::AMD {
+
+// -----------------------------------------------------------------------------
+// Pointer canonicalizer utility class
+// -----------------------------------------------------------------------------
+// This class iterates through the argument of the `funcOp`, if the argument is
+// a pointer, starts a walk through its transitive uses to build a in-memory
+// data structure to record the current offset to that pointer. Only when the
+// pointer is really loaded/stored we materialize the base pointer with the
+// offset.
+//
+// Let's suppose that `arg0` is a pointer. The algorithm works like that:
+//
+// a) At the beginning the offset is a tensor initialized to zero, and we
+//    associate with `%arg0` a `FatPtr{basePtr=%arg0, offset=0}`. Through the
+//    algorithm `FatPtr.basePtr` represents the scalar base pointer (all the
+//    uniform updates will go into that) and `FatPtr.offset` represents the
+//    tensor offset (all the non-uniform updates will go into that)
+//
+//
+// b) Follow the pointer through the IR. When we meet:
+//    `%ptr = tt.addptr(%arg0, %offset)`
+//
+//    Isolate the uniform and the non-uniform contributions of %offset =
+//    (%u_offset, %nu_offset) and update the scalar pointer and the tensor
+//    offset
+//    ```
+//    %s_ptr = addi(%fatPoniters[ptr].basePtr, %u_offset)
+//    %t_offset = addi(%fatPoniters[ptr].offset, %nu_offset)
+//    %fatPointers[%ptr0] = FatPtr{base=%s_ptr, offset=%t_offset}
+//    ```
+// c) When we meet the `tt.load(%ptr)` or `tt.store(%ptr)` instructions,
+//    replace that instruction with:
+//    `%t_ptr = tt.splat(%fatPointers[%ptr].basePtr)
+//    `%fat_ptr = tt.addptr(%t_ptr, %fatPointers[ptr].offset)`
+//    `%data = tt.load(%fat_ptr)`
+//
+class PointerCanonicalizer {
```

Comment:
nit: why is this class in the header? Is it really meant to be used outside the PointerCanonicalizer pass? Things that are in `include/` folder are meant to be public headers

Reply:
Done